_source/_posts/2022-01-12-webassembly-on-kubernetes-with-rust.md

```diff
@@ -0,0 +1,351 @@
+
+# Enter Krustlet
+
+[Krustlet](https://krustlet.dev/) is a [Kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) written in Rust for WebAssembly workloads (written in any language). It listens for new pods assigned to it based on `tolerations` specified on the manifest and runs them. Since the default Kubernetes nodes cannot run Wasm workloads natively, you need a Kubelet that can, and this is where Krustlet comes in.
+
+# Run WebAssembly workloads on Kubernetes with Krustlet
+
+Today, you will use Krustlet to run a Wasm workload written in Rust on Kubernetes.
+
+**Prerequisites**
+
+- [Docker](https://docs.docker.com/engine/install/)
+- [kubectl](https://kubernetes.io/docs/tasks/tools/)
+- [kind](https://kind.sigs.k8s.io/docs/user/quick-start/#installation) or another local kubernetes distribution
+- [Rust toolkit](https://www.rust-lang.org/learn/get-started) (Includes rustup, rustc, and cargo)
+
+## Preparing the cluster
+
+First, you need to prepare a cluster and install Krustlet on the cluster to run WebAssembly on it. I'm using [kind](https://kind.sigs.k8s.io/) to run a local Kubernetes cluster, you can also use [MiniKube](https://docs.krustlet.dev/howto/krustlet-on-minikube/), [MicroK8s](https://docs.krustlet.dev/howto/krustlet-on-microk8s/), or [another Kubernetes distribution](https://docs.krustlet.dev/intro/quickstart/).
+
+The first step is to create a cluster with the below command:
+
+```bash
+kind create cluster
+```
+
+Now you need to bootstrap Krustlet, and you will need `kubectl` installed and a kubeconfig that has access to create `Secrets` in the `kube-system` namespace and can approve `CertificateSigningRequests`. You can use the [handy script](https://github.com/krustlet/krustlet/tree/main/scripts) from Krustlet to do this. So download the appropriate script for your OS and run it.
+
+```bash
+curl https://raw.githubusercontent.com/krustlet/krustlet/main/scripts/bootstrap.sh | /bin/bash
+```
+
+Now you can install and run Krustlet.
+
+Download a binary release from the [release page](https://github.com/krustlet/krustlet/releases) and run it. I'm using the Linux version.
+
+```bash
+# Download for Linux
+curl -O https://krustlet.blob.core.windows.net/releases/krustlet-v1.0.0-alpha.1-linux-amd64.tar.gz
+tar -xzf krustlet-v1.0.0-alpha.1-linux-amd64.tar.gz
+# Install
+KUBECONFIG=~/.krustlet/config/kubeconfig \
+  ./krustlet-wasi \
+  --node-ip=172.17.0.1 \
+  --node-name=krustlet \
+  --bootstrap-file=${HOME}/.krustlet/config/bootstrap.conf
+```
+
+**Note**: If you use Docker for Mac, the node-ip would be different. Follow [the instructions from the Krustlet docs](https://docs.krustlet.dev/howto/krustlet-on-kind/#special-note-docker-desktop-for-mac) to figure out the IP.
```

Comment:
```suggestion
**Note**: If you use Docker for Mac, the node-ip would be different. Follow [the instructions from the Krustlet docs](https://docs.krustlet.dev/howto/krustlet-on-kind/#special-note-docker-desktop-for-mac) to figure out the IP. If you get the error "krustlet-wasi cannot be opened because the developer cannot be verified", you can allow it with the `allow anyway` button found at `System Preferences` / `Security & Privacy` / `General`.
```

Reply:
FWIW I had to do the same thing again later in order to run the wasm-to-oci binary

Reply:
Thanks I didn't know this